### DISCUSSION TITLE
Обсуждение

### DISCUSSION THREAD

**Discussion by Chimildic** [Chimildic/gooex#1] [Беседка]
Тема для обсуждения любых вопросов по gooex на первое время.

[Настройка и использование](https://chimildic.github.io/gooex)

**Comment by Bedmit** [comment]
Пока только начал изучать. 
Первый вопрос - как gooex согласуется с [предыдущим решением](https://github.com/Chimildic/goofy/discussions/176)? 
На сколько я понял, как минимум использует тот же токен. Ради эксперимента проверил - работают параллельно оба. 
Я так понимаю, что первое решение было просто временным и его есть смысл просто удалить. 

Второе. Какие функции (вернее их аналоги Goofy) можно использовать в Gooex? Только те что описаны в примерах в описании использования?

И третье, чисто практическое. Вроде всю настройку сделал. Даже на всякий случай получил токен заново. Проверил запустив все представленные в примерах функции. Все отлично отработало, кроме импорта из ЯМ в Споти. Постоянно переходит в паузу. Из отладки понял, что это на этапе строки `let spotifyTracks = Search.multisearchTracks(strTracks);`  Предыдущие отработаны успешно.

**Reply by Chimildic** [reply]
1. Предыдущее решение можно назвать пробой и развиваться не будет. gooex полностью его заменяет.
2. Документации по сути нет. В вики описано несколько вариантов получения треков с плейлистов. Функций больше, все доступное выпадает в подсказке после `gooex.`

   ![image](https://user-images.githubusercontent.com/40138097/163331306-4be60cc8-8c52-443d-a1b9-49d2b56316cd.png)

В целом, принцип работы схож с goofy. Есть отличие в структуре и ответах. Например, вместо `Source.getTracks` указывается явный путь `gooex.Playlist.getTracks`, что нужно получить треки именно плейлиста.

3. Вероятно проблема в количестве треков:
   - У Яндекс.Музыки есть импорт, который забирает строку со всеми названиями и ищет их на своей стороне - один запрос.
   - У Spotify такой возможности нет. Поэтому каждый трек ищем отдельно мы. Отсюда большое количество запросов. И как следствие паузы из-за "спама".

**Reply by Bedmit** [reply]
> Вероятно проблема в количестве треков

Да, действительно, 10000 лайков не переваривается. Сократил до 10 и все отлично перенеслось. 
Продолжим изучение )

**Comment by efremych** [comment]
<img width="45%" style="margin: 0" src="https://user-images.githubusercontent.com/76212904/163400779-21f7780c-cfe0-420b-9d5f-c8351081af14.png"></img>

Надеюсь я ничего не подключил) Забыл что Яндекс для людей))) Ввожу код из приложения Яндекс ключ.
Видимо, не судьба

и тут одному богу известно что за пароль имеется в виду

<img width="45%" style="margin: 0" src="https://user-images.githubusercontent.com/76212904/163401960-37c2142c-6837-4f01-8479-bab5b4239ddf.png"></img>

**Reply by Chimildic** [reply]
У меня без проблем и вопросов проходит:
Шаг 1. Вводим телефон, получаем код в смс. Вводим обычный пароль и код из смс.
Шаг 2. Придумываем пин-код.
Шаг 3. Скачиваем Яндекс.Ключ, жмм добавить аккаунт, наводим камеру на QR-код. Вводим пин-код на телефоне.
Шаг 4. Вводим одноразовый код из Яндекс.Ключа (доступен после ввода пин-кода).

**Reply by efremych** [reply]
Ну, вот, какой-то "старый пароль" просит ввести после Шага 4... Ладно, может, потом как нибудь еще на одну попытку надумаю (дважды уже скачивал приложение и удалял).

**Reply by Fridiero** [reply]
Старый пароль = старый обычный пароль от яндекса :)

**Comment by Bedmit** [comment]
Вопрос наверное на стыке Gooex и Goofy.
Хочу в Spotify пролайкать последние лайки из ЯМ. Для проверки еще и в отдельный плейлист вывести.
```
function importToSpotify() {
  let yandexLikes = gooex.Like.getLikedTracks();
  Selector.keepFirst(yandexLikes, 10);

  let strTracks = yandexTracksToString_(yandexLikes);
  Library.saveFavoriteTracks(yandexLikes);
  
  let spotifyTracks = Search.multisearchTracks(strTracks);
  Playlist.saveAsNew({
    name: 'Лайки из Яндекс.Музыки',
    tracks: spotifyTracks,
  });
}

function yandexTracksToString_(tracks) {
  return tracks.map(t => {
    t = t.track || t;
    return `${t.artists[0].name} ${t.title}`.clearName();
  });
}
```
В результате плейлист появляется, но лайки не проставлены.

Попутно вопросы. 
1. В треках из ЯМ есть информация о дне добавления? В данном случае взял 10 последних треков, т.к. фильтр получения треков добавленных сегодня не сработал.
2. Возможно ли постановка лайка в ЯМ? Если не сейчас, то в планах.
3. Возможно ли получение информации о текущем треке ЯМ? Ну и вообще что-либо из модуля Player.

**Reply by Bedmit** [reply]
Еще на некоторые треки (я так понимаю, которые не нашел) выдает ошибку 
```
Номер: 400
Адрес: https://api.spotify.com/v1/me/tracks
Текст ответа: {
  "error" : {
    "status" : 400,
    "message" : "Bad ID"
  }
} 
```
Может ее в виде инфо сделать, а не в виде ошибки?

**Reply by Chimildic** [reply]
> Хочу в Spotify пролайкать последние лайки из ЯМ

`yandexLikes` содержит лайки в формате Яндекс.Музыки. Их нельзя дать `Library.saveFavoriteTracks`, который понимает только Spotify формат треков. Для простоты можно разбить код на 3 блока: gooex, импорт, goofy.
```js
  // Только gooex -> 
  let yandexLikes = gooex.Like.getLikedTracks().sliceFirst(10);

  // <- Только gooex

  // Импорт в Spotify
  let strTracks = yandexTracksToString_(yandexLikes);
  let spotifyTracks = Search.multisearchTracks(strTracks);

  // Только goofy-> 
  Library.saveFavoriteTracks(spotifyTracks);
  Playlist.saveAsNew({
    name: 'Лайки из Яндекс.Музыки',
    tracks: spotifyTracks,
  });
  // <- Только goofy
```

> Bad ID

Ошибка связана с тем, что у треков `yandexLikes` есть поле _id_, но оно не имеет ничего общего с _id_ на стороне Spotify. Поэтому формально `Library.saveFavoriteTracks` видит и собирает _id_ в запрос, но Spotify не понимает что ему дали.

> о дне добавления

Видел некоторые поля с датами. Пока не трогал, фильтров нет. 

> постановка лайка в ЯМ

Да, работает сейчас: `gooex.Wrapper.Likes.addTracks`. Если писать вручную, попутно увидим какие еще функции есть.

> информации о текущем треке ЯМ

Насколько я понял у ЯМ нет API функций типа `play`, `pause`, `next`. Но есть что-то с очередями (получение, создание, смещение). 
Видимо текущий трек надо поиска там. Пока не разбирался.

**Reply by Bedmit** [reply]
> работает сейчас: `gooex.Wrapper.Likes.addTracks`

И вот это не понял ка работает. Хотел лайкнуть массив треков, просто вписал его как аргумент `gooex.Wrapper.Likes.addTracks(yandexTracks)` но не сработало

**Reply by Bedmit** [reply]
Вообще, если не сложно, можно готовый скрипт? Нужно пролайкать в ЯМ лайки из Spotify, а в Spotify лайки из ЯМ. В идеале за какой-то срок (например текущий день), но можно и просто какое-то количество последних (допустим 50). Только именно непролайканные треки. Если лайкать все, то они перенесутся вверх в списке лайков, а это не нужно. 
Мне кажется такой скрипт и в примерах будет полезен.

**Comment by Bedmit** [comment]
Чего-то у меня сегодня все что работало, выдает 
```
TypeError: Cannot read property 'hasOwnProperty' of undefined
_build	@ Gooex.gs:73
RequestBuilder.get	@ Gooex.gs:49
getProgress	@ Gooex.gs:904
importTracks	@ Gooex.gs:784
ExportNashe	@ ExportYndex.gs:5
```
Думал, может обновление токена поможет, но там тоже ошибка `ReferenceError: Yandex is not defined`

**Reply by Chimildic** [reply]
`Yandex` это модуль из [первого решения](https://github.com/Chimildic/goofy/discussions/176). Судя по ошибке - оно было удалено. 
Токен получаем в gooex так:
```js
function generateYandexToken() {
  gooex.Auth.generateTokenByCredentials('login', 'code');
}
```

Первая ошибка могла возникнуть если сбрасывались параметры. Что-то вроде `UserProperties.deleteAllProperties` или `Admin.reset`.

**Reply by Bedmit** [reply]
С токеном вопрос решен. Да действительно, я его пытался получить модулем первого решения. 
А вот с остальным все так же. Вроде ни каких ресетов не делал. Хотя мало ли, может чего-то такое и нажал, что не нужно. На всякий случай заново скопировал библиотеку Gooex и получил токен. Как восстановить до рабочего состояния ?

**Reply by Chimildic** [reply]
Если во время получения токена не было ошибок, то должно заработать. 
В консоле будет `access_token` после запуска следующей строки? Сюда копировать не надо.
```js
console.log(UserProperties.getProperty('yandex'))
```

**Reply by Bedmit** [reply]
Да, `Информация	{"access_token":"` и т.д.

**Reply by Chimildic** [reply]
Пример функции, где ошибка `Cannot read property 'hasOwnProperty' of undefined`?

**Reply by Bedmit** [reply]
```
function importToYandex() {
  let recentTracks = RecentTracks.get(20);
  let strTracks = spotifyTracksToString_(recentTracks);

  let yandexTracks = gooex.Importer.importTracks(strTracks);
  gooex.Playlist.saveAsNew({
    name: 'История из Spotify',
    tracks: yandexTracks,
  });
}

function spotifyTracksToString_(tracks) {
  return tracks.map(t => `${t.artists[0].name} ${t.name}`.formatName()).join('\r\n');
}
```

**Reply by Chimildic** [reply]
У меня ошибка не воспроизводится. 
Может где-то затесался кусок кода вне тела функции? Или дубликаты по именам функции.

**Reply by Bedmit** [reply]
Да вроде все проверил. Даже не знаю куда копать. Может есть смысл через TeamViever?

**Reply by Chimildic** [reply]
Можно свежий лог ошибки? И одна ли она.

**Reply by Bedmit** [reply]
```
Ошибка	
TypeError: Cannot read property 'hasOwnProperty' of undefined
_build	@ Gooex.gs:73
RequestBuilder.get	@ Gooex.gs:49
getProgress	@ Gooex.gs:864
importTracks	@ Gooex.gs:744
importToYandex	@ TestYandex.gs:27
```

**Reply by Chimildic** [reply]
Что если обновить до [сегодняшнего кода](https://github.com/Chimildic/gooex/blob/master/src/library.js)? Токен менять не надо.

**Reply by Bedmit** [reply]
Обновил. Та же ошибка.

**Reply by Chimildic** [reply]
Можно данные team viewer в сообщения 4pda

**Reply by Chimildic** [reply]
В Apps Script есть актуальная функция [PropertiesService.getUserProperties](https://developers.google.com/apps-script/reference/properties/properties-service?hl=en) и устаревший класс [UserProperties](https://developers.google.com/apps-script/reference/properties/user-properties?hl=en). В goofy одноименная константа `UserProperties`. Я так понимаю, когда файл _library_ от goofy первый в списке, происходит переприсвоение и все, кто использует это имя в последующих файлах ссылаются на goofy констану. У меня так и происходит, поэтому не было ошибки. 

Код gooex инициализировался раньше, чем goofy (по порядку файлов). Поэтому gooex ловил устаревший `UserProperties` с отдельным хранилищем параметров. При этом когда мы запускаем консоль лог `UserProperties.getProperty`, вся инициализация прошла, т.е. goofy подменил устаревший класс. Поэтому в актуальном хранилище сохранен/виден токен. Но его не видит gooex.

Вероятно между корректной работой gooex и ошибкой менялась сортировка файлов. 
Обновил [код 2022.04.16](https://github.com/Chimildic/gooex/blob/master/src/library.js). Теперь явно берется актуальное хранилище, чтобы уйти от устаревшего класса и зависимости от сортировки.

**Reply by Bedmit** [reply]
Да, вчера поменял сортировку скриптов. Сделал в алфавитном порядке. Не подумал, что это может (вернее теперь уже могло ) иметь значение. 
Теперь все ок. Причём корректно сработал скрипт экспорта в ЯМ по ночному триггеру. А библиотеку Gooex я обновил только утром. Я так понимаю, у меня на ПК все правки были сделаны ещё вчера. 
Правда теперь ещё в при выполнении в консоли куча информации выдаётся ) Но не напрягает.
Спасибо за решение!

**Reply by Chimildic** [reply]
Делал `console.log` в разных местах. Потому что отладка не даёт смотреть до инициализации. Просто эти строчки удалить надо.

**Comment by Bedmit** [comment]
Ни как не получается самостоятельно решить [эту задачу](https://github.com/Chimildic/gooex/discussions/1#discussioncomment-2584517)
Пробовал из готовых скриптов сделать перенос лайков из ЯМ в Споти
```
function ImportLike() {
  let yandexLikes = gooex.Like.getLikedTracks();
  Selector.keepFirst(yandexLikes, 50);

  let strTracks = yandexTracksToString_(yandexLikes);

  Library.checkFavoriteTracks(strTracks);
  if (!strTracks[0].isFavorite) {
    Library.saveFavoriteTracks(strTracks);
  }
}

function yandexTracksToString_(tracks) {
  return tracks.map(t => {
    t = t.track || t;
    return `${t.artists[0].name} ${t.title}`.clearName();
  });
}
```
Пробовал из Споти в ЯМ
```
function ExportLike() {
  let recentTracks = Source.getSavedTracks(50);
  let strTracks = spotifyTracksToString_(recentTracks);

  let yandexTracks = gooex.Importer.importTracks(strTracks);
  gooex.Wrapper.Likes.addTracks(yandexTracks);
  }

function spotifyTracksToString_(tracks) {
  return tracks.map(t => `${t.artists[0].name} ${t.name}`.formatName()).join('\r\n');
}
```

**Reply by Chimildic** [reply]
Не видел сообщения с той задачей. 

1. `ExportLike`, баг на стороне gooex. Яндекс по-разному данные дает. То массив объектом, то массив строк. Функция-конвертер еще не все случае отрабатывает правильно.
2. `ImportLike`, баг в самой функции. `strTracks` содержит просто строки. Нужно сначала [найти треки в Spotify](https://chimildic.github.io/gooex/usage#экспорт-из-яндекса-в-spotify) `Search.multisearchTracks` и только потом давать их модулю `Library`.

**Reply by Bedmit** [reply]
С первым, я так понимаю вопрос пока не решить?

А со вторым, сделал так 
```
function ImportLike() {
  let yandexLikes = gooex.Like.getLikedTracks();
  Selector.keepFirst(yandexLikes, 10);

  let strTracks = yandexTracksToString_(yandexLikes);
  let spotifyTracks = Search.multisearchTracks(strTracks);

  Library.checkFavoriteTracks(spotifyTracks);
  if (!spotifyTracks[0].isFavorite) {
    Library.saveFavoriteTracks(spotifyTracks);
  }
  console.log(spotifyTracks);
}

function yandexTracksToString_(tracks) {
  return tracks.map(t => {
    t = t.track || t;
    return `${t.artists[0].name} ${t.title}`.clearName();
  });
}
Но какая то ерунда получилась. Перезаписались все 10 последних лайков, поломав очередь. В консоли вообще, всегда последние два лайка.
```

**Reply by Chimildic** [reply]
Изменил код gooex, первое должно сработать. 
Про второе не понял пояснение.

**Reply by Bedmit** [reply]
Да, первое теперь отлично работает.
А со вторым, работает, но не совсем так как хотелось. Нужно, чтобы лайкались только нелайкнутые ранее треки. А на самом деле заново перелайкивается весь массив. Это нарушает важный для остальных скрпиптов порядок любимых треков. 
С консолью вообще не понятно. Просто отображает два последних лайка. Но она мне не нужна, это я так, для контроля при настройке скрипта делал.

**Reply by Chimildic** [reply]
Как раньше узнавали, Spotify перелайкивает когда даешь дубль трека. 
Сейчас проблема в условии. Не нужно смотреть первый лайк массива, надо отфильтровать. Есть в [примере](https://chimildic.github.io/goofy/#/reference/library?id=checkfavoritetracks).
```js
Library.checkFavoriteTracks(spotifyTracks);
Library.saveFavoriteTracks(spotifyTracks.filter(t => !t.isFavorite));
```

При этом также может возникнуть [проблема сортировки](https://github.com/Chimildic/goofy/discussions/178) из-за того, что треки отправляются пачками.

По поводу консоли, объект трека очень большой. Поэтому текст отрезается и видно постоянно одно и тоже. 
Можно отформатировать и выводить только названия.
```js
console.log(spotifyTracks.map(t => `${t.artists[0].name} - ${t.name}`).join('\n'))
```

**Reply by Bedmit** [reply]
> Как раньше узнавали, Spotify перелайкивает когда даешь дубль трека.

Это я пытался прикрутить кусок из другого скрипта
```
if (!newLike[0].isFavorite) {
    Library.saveFavoriteTracks(newLike);
    Cache.append('SavedTracks.json', newLike, 'begin');
    Selector.keepFirst(newLike, 1);
    Library.saveFavoriteTracks(newLike);
  }
```


Сейчас сделал с учетом вышесказанного так
```
function ImportLike() {
  let yandexLikes = gooex.Like.getLikedTracks();
  Selector.keepFirst(yandexLikes, 10);

  let strTracks = yandexTracksToString_(yandexLikes);
  let spotifyTracks = Search.multisearchTracks(strTracks);

  Library.checkFavoriteTracks(spotifyTracks);
  Library.saveFavoriteTracks(spotifyTracks.filter(t => !t.isFavorite));

  Playlist.saveAsNew({
    name: 'Новые лайки из Яндекс.Музыки',
    tracks: spotifyTracks,
  });

  console.log(spotifyTracks.map(t => `${t.artists[0].name} - ${t.name}`).join('\n'))
}

function yandexTracksToString_(tracks) {
  return tracks.map(t => {
    t = t.track || t;
    return `${t.artists[0].name} ${t.title}`.clearName();
  });
}

```
Для наглядности даже создаю плейлист. Но в итоге и в консоли, и в плейлисте просто последние 10 лайков, т.е. весь массив. А нужно, чтобы были только те лайки, которые появились после предыдущей проверки, т.е. только новые. Очередность треков в пачке этих новых уже не важна.

**Reply by Bedmit** [reply]
Забыл, как делать код цветным )

**Reply by Chimildic** [reply]
`spotifyTracks` не изменяется, поэтому на выходе всегда 10. Но на лайк отправляются только новые из-за фильтра (там создается новый массив-копия). 
Можно переписать так
```js
  Library.checkFavoriteTracks(spotifyTracks);
  spotifyTracks = spotifyTracks.filter(t => !t.isFavorite)
  Library.saveFavoriteTracks(spotifyTracks);
```

Чтобы сделать код цветным, к первой тройке символов добавляется язык js, нижнее не меняется

![image](https://user-images.githubusercontent.com/40138097/164526108-af70f45d-202f-4958-ad99-42aab7ef2812.png)

**Reply by Bedmit** [reply]
Да, да, уже и сам дошел до исправлений в коде. Теперь все ок. Конечный вариант скриптов
```js
function ExportLike() {
  let recentTracks = Source.getSavedTracks(50);
  Filter.rangeDateRel(recentTracks, 1, 1);
  let strTracks = spotifyTracksToString_(recentTracks);

  let yandexTracks = gooex.Importer.importTracks(strTracks);
  gooex.Wrapper.Likes.addTracks(yandexTracks);
}

function spotifyTracksToString_(tracks) {
  return tracks.map(t => `${t.artists[0].name} ${t.name}`.formatName()).join('\r\n');
}


function ImportLike() {
  let yandexLikes = gooex.Like.getLikedTracks();
  Selector.keepFirst(yandexLikes, 50);

  let strTracks = yandexTracksToString_(yandexLikes);
  let spotifyTracks = Search.multisearchTracks(strTracks);

  Library.checkFavoriteTracks(spotifyTracks);
  spotifyTracks = spotifyTracks.filter(t => !t.isFavorite);
  Library.saveFavoriteTracks(spotifyTracks);
}

function yandexTracksToString_(tracks) {
  return tracks.map(t => {
    t = t.track || t;
    return `${t.artists[0].name} ${t.title}`.clearName();
  });
}

```
Потестирую пару дней. Спасибо за помощь

**Comment by Bedmit** [comment]
Что-то у меня после вчерашней правки Gooex засбоили скрипты. Хотя конечно может это и ни как не связано.
Ошибки такого типа:
```js
TypeError: Cannot read property 'map' of undefined
getTrackKeys	@ Library.gs:1203
(функция без названия)	@ Library.gs:1183
removeTracks	@ Library.gs:1183
(функция без названия)	@ Service.gs:6
Service	@ Service.gs:5
```

```js
TypeError: Cannot read property 'map' of undefined
getTrackKeys	@ Library.gs:1203
(функция без названия)	@ Library.gs:1183
removeTracks	@ Library.gs:1183
BigSmartify	@ BigSmartify.gs:19
```

```js
TypeError: Cannot read property 'map' of undefined
getTrackKeys	@ Library.gs:1203
(функция без названия)	@ Library.gs:1183
removeTracks	@ Library.gs:1183
RadioTask	@ Radio.gs:51
```

О, только заметил. Началось сегодня в 1.14 с ошибки runTasks_
```js
TypeError: Cannot read property 'map' of undefined
    at getTrackKeys(Library:1203:29)
    at [unknown function](Library:1183:43)
    at removeTracks(Library:1183:31)
    at appendSavedTracks(Library:41:16)
    at runOnceAfter(Library:3455:13)
    at runTasks_(Library:30:36)
```

**Reply by Bedmit** [reply]
Скрипты не редактировал. С Gooex они не связаны. На всякий обновил библиотеку Library и даже передвинул файл выше gooex

**Reply by Chimildic** [reply]
Предположу, что в кэш лайков попал некорректный трек после очередного `appendSavedTracks`. Потому что `getTrackKeys` спотыкается об несуществующий массив исполнителей у трека `track.artists.map`.

Консоль что-нибудь покажет?
```js
let tracks = Cache.read('SavedTracks.json');
console.log(tracks.filter(t => !t.artists));
```

**Reply by Bedmit** [reply]
> Консоль что-нибудь покажет?

```
'nirvana smells like teen spirit',
  'бобры bu ga ga',
  'sting shape of my heart',
  'linkin park numb',
  'imagine dragons believer',
  'the xx intro',
  'queen another one bites the dust',
  'imagine dragons natural',
  'kongos hey i don t know',
  'najoua belyzel viola en duo avec marc lavoine album',
  'бобры bu ga ga',
  'sting shape of my heart',
  'linkin park numb',
  'imagine dragons believer',
  'the xx intro',
  'queen another one bites the dust',
  'imagine dragons natural',
  'kongos hey i don t know',
  'najoua belyzel viola en duo avec marc lavoine album',
  'aram bedrosian the mists',
 
```

Такое чувство, что ЯМ к названию артиста подобавлял название трека. Это те из треков, которые вчера через импорт лайков добавились.

**Reply by Bedmit** [reply]
И в файле на диске так и лежит одним куском, как будто это один трек. Я так понимаю, этот кусок отрезать нафиг?

**Reply by Chimildic** [reply]
Это просто строчки, не объекты Spotify, они не должны были попасть в файл _SavedTracks.json_.
Видимо некорректно внедрены функции импорта.

**Reply by Bedmit** [reply]
Пытался понять причину. Скорее всего я при самостоятельных попытках вчера так записал в файл `SavedTracks.json`.
Сперва откатил его на вчерашний день, а потом вообще решил перезаписать в него заново все лайки. В итоге, вроде все работает.

**Comment by Fridiero** [comment]
Привет! Спасибо большое за реализацию идеи кастомных плейлистов! 
Такой вопрос (я не разработчик, предупреждаю сразу :D красивых строчек кода тут не будет :)). 
Попробовал переносить "Радар новинок" из спотифая в ЯМ, но некоторые треки он меняет на похожие от исполнителя. Понятно, что в связи с, новая музыка будет достаточно редка на ЯМ, но можно ли добавлять название и исполнителя строкой в плейлист "как есть" в спотифае, пускай она даже будет недоступна на ЯМ?

Как пример ошибки при переносе:
spotify:playlist:37i9dQZEVXbhTUhQwAYstb
Второй трек - spotify:track:4AjiA5HkJVy59369f0Jjlk
в ЯМ пришел - https://music.yandex.ru/users/free.dan.man/playlists/1024
Трек - https://music.yandex.ru/album/9279878/track/60198855

То есть исполнитель основной тот же, но трек совершенно другой.

**Reply by Chimildic** [reply]
Сейчас используется импорт на стороне Яндекса. Он берет от нас названия и по своему усмотрению ищет треки. 
На практике то, что не удается найти, заменяется чем-то похожим. И как правило эта схожесть в каком-то слове. Иногда другая версия исполнения (ремикс, караоке). 

Просто недоступный трек нам не подставить. Хотел сделать простое сравнение после импорта. Если название отличается, то убирать такие треки. Тогда не будет левых, выделяющихся.

**Comment by washing2** [comment]
Добрый день. Большое спасибо за Gooex. Существует ли возможность сделать скрипт для радио (потока) по плейлисту ЯМ?

**Reply by Chimildic** [reply]
Некоторые API для потоков известны. Но с Apps Script ими управлять практически невозможно или крайне неудобно. Больше актуально для самописного приложения или хотя бы веб-формы. Ни того, ни другого нет.

Что конкретно вы хотите сделать? Зачем нужен именно поток. Есть например расширение [YaMuTools](https://chrome.google.com/webstore/detail/dgjneghdfaeajjemeklgmbojeeplehah), которое может создать похожий плейлист по данным от Last.fm.

**Reply by washing2** [reply]
Есть у меня плейлист с музыкой определенного направления. Хотелось бы, чтобы в автоматическом режиме ежедневно создавался похожий плейлист. В YaMuTools это реализовано, но для обновления листа, нужно каждый раз туда заходить. Может в YaMuTools такую функцию добавить?

**Reply by washing2** [reply]
В ЯМ эта функция есть, но в режиме реального времени и там получается лист с лайками и прослушанными композициями.

**Reply by Chimildic** [reply]
YaMuTools существует только локально в браузере и только на страницах Яндекс.Музыки. Поэтому автообновление без захода хотя бы в браузер невозможно. Я может быть попробую что-то сделать в виде приложения. Но думаю это далеко не скоро. 

С gooex другая история. Пусть с потоками сложно, но с рекомендациями для плейлиста - все просто. Добавил [пример кода](https://github.com/Chimildic/gooex/blob/master/examples/recommendation.js) для такого сбора. Вам остается добавить триггер и плейлист будет обновляться по расписанию. При желании, можно прикрутить [веб-интерфейс](https://github.com/Chimildic/goofy/discussions/9) (он больше под goofy и Spotify, но принцип схожий) и обновлять в любое время самому.

Уточните если не знаете что такое триггер для функции.

**Reply by washing2** [reply]
Спасибо. Ваши коды уже вовсю использую.

**Comment by washing2** [comment]
Добрый день, При запуске скриптов появляется ошибка.
```
13:08:48	Ошибка	
TypeError: Cannot read property 'hasOwnProperty' of undefined
_build	@ library.gs:74
RequestBuilder.get	@ library.gs:50
get	@ library.gs:931
getLikedTracks	@ library.gs:924
getLikedTracks	@ library.gs:1079
examplePlaylist	@ main.gs:8
```

**Reply by Chimildic** [reply]
Пример скрипта

**Reply by washing2** [reply]
До определенного момента все работало. Из [примера](https://github.com/Chimildic/gooex/blob/master/examples/personal-tracks.js).

**Reply by Chimildic** [reply]
Вы используете goofy? Если да, обновите его код

**Reply by washing2** [reply]
Как обновить?

**Reply by washing2** [reply]
Использую gooex.

**Reply by Chimildic** [reply]
Только gooex, без goofy?

**Reply by washing2** [reply]
У меня ЯМ. Goofy использовал когда был на споти.

**Reply by Chimildic** [reply]
В каком виде установлен gooex? Отдельный проект без кода goofy (1) или в рамках проекта goofy с новым файлом gooex (2)?

**Reply by washing2** [reply]
Сейчас использую gooex. Goofy нужно удалить?

**Reply by washing2** [reply]
Когда открываю apps script, у меня там 2 проекта goofy и gooex.

**Reply by washing2** [reply]
У меня был goofy, gooex установил по вашей инструкции.

**Reply by Chimildic** [reply]
Попробуйте повторно авторизоваться
```js
// Сгенерировать токен для Яндекс.Музыки
function generateYandexToken() {
  gooex.Auth.generateTokenByCredentials('login', 'code');
}
```

**Reply by Chimildic** [reply]
> У меня был goofy, gooex установил по вашей инструкции.

Там есть развилка "два варианта настройки gooex"

**Reply by washing2** [reply]
Уже и не помню какой из них использовал.

**Reply by washing2** [reply]
Авторизовался и все заработало. Спасибо.

**Comment by washing2** [comment]
Добрый день, Триггер для функции **generateYandexToken** как должен быть установлен?

**Reply by washing2** [reply]
Или триггер для этой функции не нужен?

**Reply by Chimildic** [reply]
Триггер не нужен. Функция запускается в ручную для получения токена. 
Если пароль от аккаунта не меняется, токен действителен год. После понадобится снова запустить её.

**Reply by washing2** [reply]
Понятно, спасибо.

**Comment by F000x** [comment]
Плохо, что в документации нет примеров кода для функций. В результате совсем непонятно, как их использовать(((
Например, я хочу удалить дубликаты артистов из плейлиста.
Вот это не работает, на выходе получаю просто пустой плейлист

```
function myFunction() {
  let playlistTracks = gooex.Playlist.getTracks('1004');
  let newTracks = gooex.Filter.dedupArtists(playlistTracks);
  gooex.Playlist.saveAsNew({
    name: 'Плейлист2',
    tracks: newTracks,
    description: 'Плейлист создан с помощью gooex',
    visibility: 'private'})
    }
```

**Reply by Chimildic** [reply]
У проекта очень маленький охват. Даже по сравнению с [goofy](https://github.com/Chimildic/goofy). Отсюда отсутствие активности и подробностей в документации. Попробуйте следующий код.
```js
  let playlistTracks = gooex.Playlist.getTracks('1004').map(item => item.track);
  gooex.Filter.dedupArtists(playlistTracks);

  gooex.Playlist.saveAsNew({
    name: 'Плейлист2',
    tracks: playlistTracks,
    description: 'Плейлист создан с помощью gooex',
    visibility: 'private'
  })
```

**Reply by F000x** [reply]
Спасибо за ответ. Скопировал ваш код, запустил (перед этим обернув в функцию) и всё равно на выходе получаю пустой плейлист.

**Reply by Chimildic** [reply]
Вы уверены, что плейлист `1004` не пустой?

**Reply by F000x** [reply]
Конечно, другой непустой плейлист тоже пробовал. Вот ссылка https://music.yandex.ru/users/IssaiMalueg19931/playlists/1004

**Reply by Chimildic** [reply]
Аккаунт, с которого вы пытаетесь взять треки и аккаунт, который авторизован в gooex - один и тот же? 
Попробуйте добавить логин пользователя
```js
let playlistTracks = gooex.Playlist.getTracks('1004', 'IssaiMalueg19931').map(item => item.track);
```

**Reply by F000x** [reply]
Да, один и тот же. На том же аккаунте IssaiMalueg19931 появляется пустой плейлист 'Плейлист2'. Попробовал взять чужие плейлисты, подставив в код ID плейлиста и название чужого аккаунта, тоже получаю пустой плейлист на выходе. У меня даже тестовый скрипт с функцией examplePlaylist некорректно работает. У меня 10 лайкнутых треков. Запускаю функцию examplePlaylist и получаю пустой плейлист Случайные лайки. Попробовал заново скопировать проект в AppsScript и сгенерировать токен - не помогло, плейлисты на аккаунте создаются, но они пустые.

**Reply by Chimildic** [reply]
Очень странно. При выполнении кода есть ошибки?

**Reply by F000x** [reply]
Тоже не знаю, в чем дело. Нет ошибок. Вот скринкаст, как все происходит https://www.youtube.com/watch?v=POXM8a_YMq0

**Reply by Chimildic** [reply]
Проблема в отсутствии подписки на Яндекс.Плюс. Со стороны gooex нет никаких условий и ограничений. 
Но Яндекс.Музыка странно себя ведет с треками, когда подписки нет. Например, импорт треков не находит никаких треков или дает совершенно левые. Для добавления треков в плейлист как раз используется импорт. Почему это так работает не представляю. Возможно их баг, возможно намеренно. Учитывая неофициальное API жаловаться не приходится. 
Попробуйте промокоды с бесплатным доступом. К примеру, [отсюда](https://www.pepper.ru/coupons/plus.yandex.ru) или с [форума](https://4pda.to/forum/index.php?showtopic=1005425).

**Reply by F000x** [reply]
Спасибо! Проблема решилась. Действительно, дело было в Плюсе. Не думал, что все работает через импорт.

**Reply by Chimildic** [reply]
В рамках одной Яндекс.Музыки гоняются его внутренние id треков, но как видите тоже без Плюса не хочет работать.
